### DISCUSSION TITLE
how are environment variables set in the framework?

### DISCUSSION THREAD

**Discussion by bitterbark** [NOAA-GFDL/MDTF-diagnostics#464] [General]
As far as I know, this only applies to the NCL pods that rely on environment variables instead of using the python dictionaries. 

Can you point me to where the input settings (either for individual pods, from settings.jsonc, or the global vars from default_test.jsonc) are exported as environment variables? I expect this to happen using os.environ but the only instances are to import env vars into python.

--
My working day may not be the same as your working day. Please do not feel obliged to reply to this email outside of your normal hours.

**Comment by wrongkindofdoctor** [comment]
@bitterbark  

The environment variables are defined in the `set_pod_env_vars` methods in the [Diagnostic](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/f45de443b76947e3dbece10b0f553f3d4c743db5/src/diagnostic.py#L1036-L1092) and [MultirunDiagnostic](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/f45de443b76947e3dbece10b0f553f3d4c743db5/src/diagnostic.py#L1389-L1437) classes.


The `set_pod_env_vars` methods are called via the `pre_run_setup` methods by `setup_env_vars` in the  [SubprocessRuntimePODWrapper](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/f45de443b76947e3dbece10b0f553f3d4c743db5/src/environment_manager.py#L362-L404) and [MultirunSubprocessRuntimePODWrapper](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/f45de443b76947e3dbece10b0f553f3d4c743db5/src/environment_manager.py#L622-L732) classes

Note that the MultirunSubprocessRuntimePODWrapper also contains the `write_case_env_file` method to export the environment variables to the case_info.yaml file.

**Reply by bitterbark** [reply]
Thanks, @wrongkindofdoctor . I'm still confused, although I'm sure it is my ignorance of how to penetrate python code that appears to be circular but does actually do something. How does the run-time environment actually learn about the pod_env_vars? I'm looking for  the equivalent of invoking 
> os.environ[varname]=value

so that another program (NCL) can extract the values for itself?

If it helps to know that I am trying to communicate the information from the <write_case_env_file>, which I can read into a dict with the code you provided in the example_multicase, to my NCL program.

**Reply by wrongkindofdoctor** [reply]
@bitterbark The environment variables defined in `env_vars` are passed to the PODs via the `env` argument to the [suprocess](https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module) call that launches each POD in the [subprocessRuntimeManager](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/f45de443b76947e3dbece10b0f553f3d4c743db5/src/environment_manager.py#L533-L559). The single-run PODs can then access each environment variable with `os.environ`.
